### PR TITLE
Pretty Printer

### PR DESCRIPTION
**Pull request by anthonyholley**
finally figured out how to do this again - pretty printer incomplete from earlier

### CODE REVIEW

File: src/prettyprint.ml [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+
+(* Print Operators BRUG*)
+	
+let string_of_bin_op = function
+    Add        -> "+"
+  | Minus      -> "-"
+  | Multiply   -> "*"
+  | Divide     -> "/" 
+  | Modulo     -> "%"
```
Comment (xvladus1) [inline src/prettyprint.ml]: this is `mod` in OCaml instead of `%`

File: src/prettyprint.ml [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+  | Bit_LShift    -> "<<"
+  | Bit_And       -> "&"
+  | Bit_Or        -> "|"
+  | Bit_Xor       -> "^"
+
+ 
+(* Print Expressions BRUG*)
+  
+let rec string_of_expr = function
+    Binop(e1, o, e2)	-> "(" ^ (string_of_expr e1) ^ " " ^
```
Comment (vlad-scherbich) [inline src/prettyprint.ml]: I think `"("` at the beginning and `")"` at the end aren't necessary.

File: src/prettyprint.ml [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+                                 (string_of_bin_op o) ^ " " ^
+							     (string_of_expr e2) ^ ")"
+								 
+  | Bitop(i1, o, i2)	-> "(" ^ (string_of_int i1) ^ " " ^
```
Comment (vlad-scherbich) [inline src/prettyprint.ml]: same as above

File: src/prettyprint.ml [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+							   (string_of_actor a) ^ ")"
+  | Noexpr              -> ""
+  | _                   -> "brug"
+
+ 
+(* Print Statements BRUG*)
+
+let rec string_of_stmt = function
+    Block(stmts)        -> "{\n" ^ String.concat "\n"
+	                         (List.map string_of_stmt stmts) ^ "}\n"
```
Comment (vlad-scherbich) [inline src/prettyprint.ml]: should we add a `\n` before closing `}`?

File: src/prettyprint.ml [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+  | Expr(expr)          -> string_of_expr expr ^ ";\n";
+  | Return(expr)        -> "return " ^ string_of_expr expr ^ ";\n";
+  | If(e, s, Block([])) -> "if (" ^ string_of_expr e ^ ")\n" ^ string_of_stmt s
```
Comment (vlad-scherbich) [inline src/prettyprint.ml]: Are we missing LBRACE and RBRACE around `string_of_stmt s` ?

File: src/prettyprint.ml [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+  | If(e, s1, s2)       -> "if (" ^ string_of_expr e ^ ")\n" ^
```
Comment (vlad-scherbich) [inline src/prettyprint.ml]: same as above

File: src/prettyprint.ml [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+                             string_of_stmt s1 ^ "else\n" ^ string_of_stmt s2
+  | For(fi, e2, e3, st) -> ("for (" ^ (string_of_for_init fi) ^ " ; " ^
```
Comment (vlad-scherbich) [inline src/prettyprint.ml]: same as above

File: src/prettyprint.ml [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+                             (string_of_expr e2) ^ " ; " ^ (string_of_expr e3)
+							   ^ ") " ^ (string_of_stmt st))
+  | While(e, st)        -> ("while (" ^ (string_of_expr e) ^ ") " ^
```
Comment (vlad-scherbich) [inline src/prettyprint.ml]: same as above

File: src/prettyprint.ml [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+  "actor " ^ actor.name ^ "(" ^ String.concat ", " (List.map string_of_formal
+    func.formals) ^ ") {\n  " ^ String.concat "\n  " (List.map string_of_stmt
+	  actor.body) ^ "\n\n  " ^ String.concat "\n\n  " (List.map string_of_func
+	    actor.functions) ^ "\n\n  " ^ "receive {\n    " ^ String.concat "\n    "
+		  (List.map string_of_pattern actor.receive) ^ "\n  }\n}"
+
+
+(* Print Lambda Functions BRUG*)
+
+(*let string_of_lambda lambda =
```
Comment (vlad-scherbich) [inline src/prettyprint.ml]: lambda is now defined in AST

File: src/prettyprint.ml [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+(* Words That Do Stuff *) 
+  
+let _ =
+  let lexbuf = Lexing.from_channel stdin in
+  let program = Parser.program Scanner.token lexbuf in
+  let result = string_of_program program in
+  print_endline result
+  
+  
+
```
Comment (vlad-scherbich) [inline src/prettyprint.ml]: nit: please add a newline (carriage return) at the end of this file

**Comment by Howon** [comment]
Is this functional?

File: src/prettyprint.ml (hunk 2 above)
Comment (jibben) [reply src/prettyprint.ml]: I think we have these "(" and ")" to enable us to interpret the order of operations as seen by the compiler. Yes its not strictly necessary, but its helpful

**Pull request closed without merging**